File: app/lib/security.ts
import { NextRequest, NextResponse } from 'next/server'
import { SECURITY_CONFIG } from '../config/app'

/**
 * 🛡️ Комплексная система защиты от DDoS и атак
 */

// Rate limiting storage (в продакшене используйте Redis)
interface RateLimitEntry {
  count: number
  resetTime: number
  blocked: boolean
  blockUntil?: number
}

const rateLimitStore = new Map<string, RateLimitEntry>()

// Очистка старых записей (из конфигурации)
setInterval(() => {
  const now = Date.now()
  for (const [key, entry] of rateLimitStore.entries()) {
    if (entry.resetTime < now && (!entry.blockUntil || entry.blockUntil < now)) {
      rateLimitStore.delete(key)
    }
  }
}, SECURITY_CONFIG.RATE_LIMIT_CLEANUP_INTERVAL_MS)

/**
 * Получает IP адрес из запроса (с учетом Cloudflare и прокси)
 */
export function getClientIP(request: NextRequest): string {
  // Cloudflare передает реальный IP в заголовке
  const cfIP = request.headers.get('cf-connecting-ip')
  if (cfIP) return cfIP

  // X-Forwarded-For (может содержать несколько IP)
  const xForwardedFor = request.headers.get('x-forwarded-for')
  if (xForwardedFor) {
    const ips = xForwardedFor.split(',').map(ip => ip.trim())
    return ips[0] // Первый IP - это оригинальный клиент
  }

  // X-Real-IP
  const xRealIP = request.headers.get('x-real-ip')
  if (xRealIP) return xRealIP

  // Fallback
  return request.ip || 'unknown'
}

/**
 * Валидация входных данных для защиты от инъекций
 */
export function sanitizeInput(input: any): any {
  if (typeof input === 'string') {
    // Удаляем потенциально опасные символы
    return input
      .replace(/[<>]/g, '') // Удаляем HTML теги
      .replace(/['";\\]/g, '') // Удаляем SQL инъекции
      .trim()
  }
  
  if (Array.isArray(input)) {
    return input.map(sanitizeInput)
  }
  
  if (input && typeof input === 'object') {
    const sanitized: any = {}
    for (const [key, value] of Object.entries(input)) {
      sanitized[key] = sanitizeInput(value)
    }
    return sanitized
  }
  
  return input
}

/**
 * Проверяет наличие SQL инъекций в строке
 */
export function containsSQLInjection(input: string): boolean {
  // Более точная проверка на SQL инъекции - только реальные паттерны атак
  const sqlPatterns = [
    // SQL команды (только если это не часть обычного текста)
    /\b(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE|UNION|SCRIPT)\s+.*(FROM|INTO|TABLE|DATABASE|WHERE)/i,
    // Комментарии SQL в контексте (-- или # в начале строки или после пробела, за которыми следует SQL-подобный текст)
    /(--|#)\s*(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE|UNION)/i,
    // Многострочные комментарии SQL
    /\/\*.*(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE|UNION).*\*\//i,
    // SQL инъекции через OR/AND с двойными равенствами
    /(\bOR\b.*=.*=)/i,
    /(\bAND\b.*=.*=)/i,
    // SQL инъекции через кавычки с OR/AND
    /('|"|`).*(\bOR\b|\bAND\b).*('|"|`)/i,
    // Попытки завершить SQL запрос точкой с запятой перед SQL командами
    /;.*(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE|UNION)/i,
  ]
  
  return sqlPatterns.some(pattern => pattern.test(input))
}

/**
 * Rate Limiter с защитой от DDoS
 */
export interface RateLimitOptions {
  windowMs?: number // Окно времени в миллисекундах (по умолчанию 1 минута)
  maxRequests?: number // Максимальное количество запросов (по умолчанию 60)
  blockDurationMs?: number // Время блокировки при превышении (по умолчанию 15 минут)
  keyGenerator?: (request: NextRequest) => string // Функция для генерации ключа
}

export function rateLimit(options: RateLimitOptions = {}) {
  const {
    windowMs = SECURITY_CONFIG.RATE_LIMIT_WINDOW_MS,
    maxRequests = SECURITY_CONFIG.RATE_LIMIT_MAX_REQUESTS,
    blockDurationMs = SECURITY_CONFIG.RATE_LIMIT_BLOCK_DURATION_MS,
    keyGenerator = (req) => `rate_limit:${getClientIP(req)}:${req.nextUrl.pathname}`,
  } = options

  return (request: NextRequest): NextResponse | null => {
    const key = keyGenerator(request)
    const now = Date.now()
    let entry = rateLimitStore.get(key)

    // Проверяем, не заблокирован ли IP
    if (entry?.blocked && entry.blockUntil && entry.blockUntil > now) {
      const remainingBlockTime = Math.ceil((entry.blockUntil - now) / 1000)
      return NextResponse.json(
        {
          error: 'Too many requests',
          message: `IP temporarily blocked. Try again in ${remainingBlockTime} seconds.`,
        },
        {
          status: 429,
          headers: {
            'Retry-After': remainingBlockTime.toString(),
            'X-RateLimit-Limit': maxRequests.toString(),
            'X-RateLimit-Remaining': '0',
            'X-RateLimit-Reset': entry.blockUntil ? new Date(entry.blockUntil).toISOString() : new Date().toISOString(),
          },
        }
      )
    }

    // Создаем новую запись или сбрасываем счетчик
    if (!entry || entry.resetTime < now) {
      entry = {
        count: 0,
        resetTime: now + windowMs,
        blocked: false,
      }
    }

    // Увеличиваем счетчик
    entry.count++

    // Проверяем превышение лимита
    if (entry.count > maxRequests) {
      entry.blocked = true
      entry.blockUntil = now + blockDurationMs
      rateLimitStore.set(key, entry)

      return NextResponse.json(
        {
          error: 'Too many requests',
          message: 'Rate limit exceeded. Your IP has been temporarily blocked.',
        },
        {
          status: 429,
          headers: {
            'Retry-After': Math.ceil(blockDurationMs / 1000).toString(),
            'X-RateLimit-Limit': maxRequests.toString(),
            'X-RateLimit-Remaining': '0',
            'X-RateLimit-Reset': new Date(entry.blockUntil).toISOString(),
          },
        }
      )
    }

    // Сохраняем обновленную запись
    rateLimitStore.set(key, entry)

    // Возвращаем null, если все в порядке (продолжаем обработку)
    return null
  }
}

File: bot_new/security.py
#!/usr/bin/env python3
"""
🛡️ Модуль защиты от DDoS и атак для Telegram ботов
"""

import time
import logging
from collections import defaultdict
from typing import Dict, Optional

logger = logging.getLogger(__name__)

_rate_limit_store: Dict[str, Dict] = defaultdict(dict)
_blocked_users: Dict[int, float] = {}


def get_user_id(update) -> Optional[int]:
    """Извлекает user_id из update"""
    if hasattr(update, 'effective_user') and update.effective_user:
        return update.effective_user.id
    if hasattr(update, 'message') and update.message and update.message.from_user:
        return update.message.from_user.id
    if hasattr(update, 'callback_query') and update.callback_query and update.callback_query.from_user:
        return update.callback_query.from_user.id
    return None

RATE_LIMIT_WINDOW = 60
MAX_REQUESTS_PER_WINDOW = 30
BLOCK_DURATION = 900


def is_user_blocked(user_id: int) -> bool:
    """Проверяет, заблокирован ли пользователь"""
    if user_id in _blocked_users:
        unblock_time = _blocked_users[user_id]
        if time.time() < unblock_time:
            return True
        else:
            del _blocked_users[user_id]
    return False


def block_user(user_id: int, duration: int = BLOCK_DURATION) -> None:
    """Блокирует пользователя на указанное время"""
    unblock_time = time.time() + duration
    _blocked_users[user_id] = unblock_time
    logger.warning(f"🚫 User {user_id} blocked for {duration} seconds")


def check_rate_limit(user_id: int):
    """
    Проверяет rate limit для пользователя
    Возвращает (is_allowed, error_message)
    """
    if is_user_blocked(user_id):
        remaining_time = int(_blocked_users[user_id] - time.time())
        return False, f"User temporarily blocked. Try again in {remaining_time} seconds."
    
    now = time.time()
    user_data = _rate_limit_store.get(user_id, {})
    
    reset_time = user_data.get('reset_time', 0)
    if reset_time < now:
        user_data = {
            'count': 0,
            'reset_time': now + RATE_LIMIT_WINDOW
        }
    
    user_data['count'] = user_data.get('count', 0) + 1
    
    if user_data['count'] > MAX_REQUESTS_PER_WINDOW:
        block_user(user_id, BLOCK_DURATION)
        logger.warning(f"🚫 Rate limit exceeded for user {user_id}. Blocked for {BLOCK_DURATION} seconds.")
        return False, "Rate limit exceeded. You have been temporarily blocked."
    
    _rate_limit_store[user_id] = user_data
    
    return True, None


def validate_input(text: Optional[str], max_length: int = 4096):
    """
    Валидирует входной текст
    Возвращает (is_valid, error_message)
    """
    if text is None:
        return True, None
    
    if len(text) > max_length:
        return False, f"Text too long. Maximum length is {max_length} characters."
    
    import re
    sql_patterns = [
        r'\b(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE|UNION|SCRIPT)\s+.*(FROM|INTO|TABLE|DATABASE|WHERE)',
        r'(--|#)\s*(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE|UNION)',
        r'/\*.*(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE|UNION).*\*/',
        r'\bOR\b.*=.*=',
        r'\bAND\b.*=.*=',
        r"('|`|\").*(\bOR\b|\bAND\b).*('|`|\")",
        r';.*(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE|UNION)',
    ]
    
    for pattern in sql_patterns:
        if re.search(pattern, text, re.IGNORECASE):
            return False, "Invalid input detected."
    
    xss_patterns = [
        r'<script',
        r'javascript:',
        r'onerror=',
        r'onload=',
        r'<iframe',
    ]
    
    for pattern in xss_patterns:
        if re.search(pattern, text, re.IGNORECASE):
            return False, "Invalid input detected."
    
    return True, None


def sanitize_input(text: str) -> str:
    """Очищает входной текст от потенциально опасных символов"""
    import html
    text = html.escape(text)
    text = text.replace("'", "").replace('"', '').replace(';', '').replace('--', '')
    return text

